### ISSUE TITLE
Fighting h@ckers with Google Recaptcha v3

### ISSUE THREAD

**Issue by bdebon** [bdebon/choiceof.dev#100]
For now, the website can easily be hacked. Someone with little knowledge can send a request with a new fake IP every time and voting 1000th times easily for the same response. This is what happened with Php vs JavaScript. A backend dev discovered the breach and decided to put 6000 votes for PHP...
Don't worry, we will catch up with these 6000 fake votes with 6001 real people's vote for javascript.

To fight this, several solutions has been proposed
- auth or auth0 : I don't want any authentication system that would break the fun and the friendly ux.
- Entering a mail: I don't want either
- Using [Google Recaptcha v3](https://developers.google.com/recaptcha/docs/v3) to generate a token than the person must attach to his request

I like this last solution but I don't have too much time to dig into how to implement that but if someone wants to do it... You're welcome!!

**Comment by kmartin91** [comment]
![image](https://user-images.githubusercontent.com/3769597/203382959-fe2a6823-414e-421c-9a37-5302e3bedbcf.png)
(recaptcha entreprise / V3 / V2)
https://www.google.com/recaptcha/about/

Don't break internet again, it might cost you a bit

**Comment by MaximeMRF** [comment]
A lot of e-commerces uses [cloudflare bot management](https://www.cloudflare.com/products/bot-management/) to block scrapers.

I think, it can be a good solution for this problem because simple post request made by a script is blocked by cloudflare.

In addition, cloudflare may also detect if you are using a browser managed by a program like puppeteer (node.js) or selenium (python).

On the other hand, I dunno how cost this solution.

**Comment by Gelocg** [comment]
Why can't we just use cloudflare ? It might do the trick

**Comment by bdebon** [comment]
I don't know, never used it. Is it compatible with the technologies we can use with the shared hosting at Hostinger ? I precise we just have PHP / MySQL / and basic nginx. No node possible.

**Comment by Snox-dev1** [comment]
It's really too easy even without scraping @MaximeMRF 
![image](https://user-images.githubusercontent.com/12135385/203650698-298ef3fa-62b8-4ec0-8fe0-44f0d8d8f0d4.png)

**Comment by Snox-dev1** [comment]
The best way is to get the voter's ip address and block him when he votes too much

**Comment by Kaporos** [comment]
Is really captcha a good Idea ? Anyone could just extract the token from his browser and put it in any script. To be sure that can't be done, a captcha should be asked for every question but it's sooo bad for ux

**Comment by Snox-dev1** [comment]
I had an idea ! 

Basically on the database you have to add a timestamp of the last vote

Then at each vote look when was the last vote of the ip then if it was less than 5 minutes ago then we block.

**Comment by bdebon** [comment]
I sounds incredible! Easy to implement and good enough to prevent the biggest abuses! Does anybody has something agains this proposition?

**Comment by bdebon** [comment]
oops no it does not work. Anyone can change it's IP for a random one. Node scripts are doing that super easily, so it's not working.

**Comment by MatteoGauthier** [comment]
A solution which does not cover all cases may be the use of [fingerprintjs](https://fingerprint.com) on the front in order to prevents browser automation.

**Comment by quentint** [comment]
The only viable solution is a backend one, since the front/JS can easily be bypassed. So any JS library sounds irrelevant to me 😉 

What could help is a PHP rate limiter, eg. https://symfony.com/doc/current/rate_limiter.html (or letting Cloudflare handle that, if technically possible).

**Comment by quentint** [comment]
> ![image](https://user-images.githubusercontent.com/3769597/203382959-fe2a6823-414e-421c-9a37-5302e3bedbcf.png) (recaptcha entreprise / V3 / V2) https://www.google.com/recaptcha/about/
> 
> Don't break internet again, it might cost you a bit

If you're looking for a cheaper/free alternative, https://www.hcaptcha.com/ is a good solution.

**Comment by bdebon** [comment]
Hey @quentint!! So nice to have you here!! I would love to have your avatar as a contributor of this project!! If you want to be the hero that will finally close the breach of this project, you are more than welcome and I can make anything that you need to implement this feature!

**Comment by quentint** [comment]
My brain has this issue running in the background. Will post when implementation details pop!

**Comment by kmartin91** [comment]
> > ![image](https://user-images.githubusercontent.com/3769597/203382959-fe2a6823-414e-421c-9a37-5302e3bedbcf.png) (recaptcha entreprise / V3 / V2) https://www.google.com/recaptcha/about/
> > Don't break internet again, it might cost you a bit
> 
> If you're looking for a cheaper/free alternative, https://www.hcaptcha.com/ is a good solution.

you need to pay if you want to use hCaptcha without captcha

**Comment by bdebon** [comment]
@quentint I think there is no better person than you to fix that, just because choixdemerde.fr was first inspired by generationwhat and you know what duo was behind this one...

**Comment by quentint** [comment]
> you need to pay if you want to use hCaptcha without captcha

Really? I think I've used it in the past without paying anything, and I can't confirm what you're saying when reading the docs 🤔

**Comment by kmartin91** [comment]
> > you need to pay if you want to use hCaptcha without captcha
> 
> Really? I think I've used it in the past without paying anything, and I can't confirm what you're saying when reading the docs 🤔

![image](https://user-images.githubusercontent.com/3769597/204010553-93a5304c-71a5-4e84-8c67-c4fc43b27484.png)

Need a pro account to have the "no captcha mode" according to https://www.hcaptcha.com/#plans . Maybe I'm wrong, I didn't dig that more

**Comment by quentint** [comment]
I think that means combining both solutions, at least that's what I found in the docs: https://docs.hcaptcha.com/invisible#invisible-vs-passive

> [Enterprise](https://www.hcaptcha.com/enterprise?utm_campaign=docs_invis) users can combine "Invisible" (no checkbox) configuration with "Passive" difficulty to avoid any user interruption.

@bdebon: Is the Cloudflare activated on Hostinger? eg. https://support.hostinger.com/en/articles/1583241-how-to-activate-and-deactivate-cloudflare

If so, there might be a way to rate limit from there. See [Configure Thresholds](https://www.cloudflare.com/en-gb/rate-limiting/?utm_source=google&utm_medium=cpc&utm_campaign=*_DG_%7c_EMEA_%7c_V2_%7c_FRANCE_%7c_G_%7c_Search_%7c_Pure_Brand&utm_content=&utm_term=cloudflare%20rate%20limiting&campaignid=18324392831&adgroupid=141199980236&creativeid=621538053286&&_bt=621538053286&_bk=cloudflare%20rate%20limiting&_bm=e&_bn=g&_bg=141199980236&_placement=&_target=&_loc=9055289&_dv=c&awsearchcpc=1&gclid=CjwKCAiA7IGcBhA8EiwAFfUDsVjFcaAPVclHhv5ZQn3s1iCdIt2CLxB-5qETCwUB8XR8RxhWLMEkxBoCOAQQAvD_BwE&gclsrc=aw.ds#:~:text=Configure%20Thresholds).

**Comment by quentint** [comment]
For now https://developers.cloudflare.com/turnstile/ looks like the way to go for me.  

Here are the ➕: 
- Free
- Unlimited
- Works with any host/domain
- Does not rely on IP or any fingerprinting
- Has a front and back API to ensure a regular user workflow

Here is the only ➖ I found:
- Currently in Beta, so it could break/disappear

Thoughts anyone?

**Comment by EliottCestSwag** [comment]
Maybe you could had a slide bar that bot can not do. When you put to much connection on Leboncoin.com, the website will ask you to fill a little puzzle with a slidebar, maybe somebody could had that type of slidebar (without the puzzle) to switch between each images without ruining the ux ?

**Comment by tdaron** [comment]
What about a JWT gained with ReCaptcha, allowing user to answer *one* time to every question (and the captcha requirement could depend of the website utilisation, if no-one is playing, no need to captcha, but if at the same time there is 1000 concurrent connections, then captcha for everyone)

**Comment by EliottCestSwag** [comment]
> What about a JWT gained with ReCaptcha, allowing user to answer _one_ time to every question (and the captcha requirement could depend of the website utilisation, if no-one is playing, no need to captcha, but if at the same time there is 1000 concurrent connections, then captcha for everyone)

Yeah but If @bdebon decide to open """officially""" the website maybe the site will detect 10.000 people at the same time and block them all or ruining the ux because of the captcha :/

**Comment by tdaron** [comment]
Yeah but it's one captcha for the first connection then not anymore for every question, and in any case its possible to do difference between bot rush or people rush i guess

**Comment by EliottCestSwag** [comment]
> Yeah but it's one captcha for the first connection then not anymore for every question, and in any case its possible to do difference between bot rush or people rush i guess

The bot IP and people IP are really different, if a backend could add a prevent against that that could be great what do u think ? @bdebon
